### ISSUE TITLE
reminder: history timing attacks

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/user.js#933]
see #448 

- [1632765](https://bugzilla.mozilla.org/show_bug.cgi?id=1632765) - just this bug will do, some of the others are access denied
- refer to Emilio's email (@earthlng, I sent you a copy months ago: ask if you want it again) on how it works: relies on other prefs combo
- test - @earthlng has some private tests for history stealing
- add to user.js
- get naked, drink beer, assign jesus was listening label (in any order)

**Comment by Thorin-Oakenpants** [comment]
So 77 flipped
```
pref("layout.css.always-repaint-on-unvisited", true); // prev: false
pref("layout.css.notify-of-unvisited", true); // prev: false
```

and we have the existing 
```
user_pref("layout.css.visited_links_enabled", false);
```

What I don't have is a history stealing PoC - @earthlng 

My assumption is that we want all three as true for it to work (I'll have to recheck what emilio said): but I'm not sure how to do that: what with ESR

**Comment by earthlng** [comment]
I'll do some testing in FF77 and report back.
ATM it looks like a recent update to RFP already thwarts my PoC even without those new mitigations in FF77

**Comment by earthlng** [comment]
Test results are in! ;)

1st of all, some recent changes to RFP always round the timings of the method I use in my PoC to only a precision of 100ms and thus make the attack pretty much impossible AFAICT.
RFP in ESR68.x doesn't have that protection yet though.

Without RFP, the 2 new prefs seem to suffer from the same problem as their last mitigation attempt in that they merely make visited links return lower timings instead of vice-versa.
They *do* slow the attack down though but the best mitigation still seems to be RFP at this point.
IDK if the repainting due to the new mitigations result in noticeable performance impacts during normal browsing.

The value of `layout.css.visited_links_enabled` doesn't seem to matter.

PoC results with the new mitigations and `layout.css.visited_links_enabled=false`
![1](https://user-images.githubusercontent.com/25846669/83646955-da5cf980-a5a3-11ea-95ac-d4277f4947f4.png)

PoC results (w same parameters) with the new mitigations and `visited_links_enabled=true`
![2](https://user-images.githubusercontent.com/25846669/83646972-dfba4400-a5a3-11ea-9d86-4a4a577ea012.png)

Visited links were mintpressnews, bugzilla and RT.

For comparison, running the PoC with the same parameters as used for the pix above but **without** the new mitigations results in average timings of ~66ms.
But to get accurate history stealing results in that case, I need to adjust the PoC parameters to achieve average timings of around 100ms (= the total PoC runtime would still be about 2 to 3 times faster than **with** the mitigations)

I could probably optimize the PoC parameters a bit to make the attack overall faster but this is just to illustrate that the new mitigations on their own don't really make history stealing impossible.

**Comment by Thorin-Oakenpants** [comment]
So .. just asking .. are you attempting to bypass any precision clamping etc by running your own timer - e.g. just run a 1ms incremental counter and refer to it?

**Comment by earthlng** [comment]
no I didn't try that

**Comment by Thorin-Oakenpants** [comment]
I sent you an email

**Comment by Thorin-Oakenpants** [comment]
@earthlng FY

- article: https://blog.lukaszolejnik.com/web-browsing-histories-are-private-personal-data-now-what/
- research pdf: https://www.usenix.org/system/files/soups2020-bird.pdf

from the article
> the 50 most popular sites only is sufficient to obtain unique fingerprints

makes sense 2^49 or whatever ... of course in the timing attack, you need exact url matches (right?) .. using the landing page of each site might be enough .. IDK: what if they redirect to other TLDs (example.com -> example.de), or use the language (which you could get via navigator: e.g. example.com/en/ loads example/de/ )...

**Comment by Thorin-Oakenpants** [comment]
[link](https://bugzilla.mozilla.org/buglist.cgi?bug_id=557579,773338,884270,1131288,1632765) 557579, 773338, 884270, 1131288 considered fixed in 1632765

That and we use RFP which has timing protection (there are timer techniques to bypass RFP in this regard).. We also clear history on close.

We currently disable link visitedness : 0805 `layout.css.visited_links_enabled` = false. This should no longer make any difference since FF77 (so ESR78 users are covered) when [1632765](https://bugzilla.mozilla.org/show_bug.cgi?id=1632765) landed
